### DISCUSSION TITLE
Conflicting peerDependencies npm@7

### DISCUSSION THREAD

**Discussion by oeduardoal** [npm/feedback#191] [General]
I believe this change will break many CIs that work with npm@latest...

In the new version of NPM (v7). By default, npm install will fail when it encounters conflicting peerDependencies ... It was not like that before ...

https://github.blog/2021-02-02-npm-7-is-now-generally-available/#peer-dependencies

the differences between them

`--legacy-peer-deps`: ignore all peerDependencies when installing, in the style of npm version 4 through version 6.

`--strict-peer-deps`: fail and abort the install process for any conflicting peerDependencies when encountered. By default, npm will only crash for peerDependencies conflicts caused by the direct dependencies of the root project.

Now, `--strict-peer-deps` is as standard...

**Comment by ljharb** [comment]
For the vast majority of my packages, nothing broke here. The only case was a plugin that tests multiple versions of eslint, and also uses a TypeScript eslint parser that doesn't support all of those versions - so I used `--legacy-peer-deps`, and everything's fine.

Any project where `npm ls` exits nonzero has an invalid dependency graph, so it's a *very* helpful thing that npm 7 surfaces this during installation.

**Reply by MylesBorins** [reply]
I'm going to mark this as the answer to this question. This was an intentional breaking change.

FWIW `--strict-peer-deps` is not the default, there are scenarios that npm can resolve without failing that will throw with `--strict-peer-deps` but pass by default in npm 7

**Comment by lemoun** [comment]
[Npm 7 breaks opencv4nodejs package installation](https://github.com/justadudewhohacks/opencv4nodejs/issues/796), with the message ' throw new Error('no OpenCV libraries found in lib dir: ' + libDir)'. We suspect it is related to peer dependencies, so we tried with `--legacy-peer-deps`, with no effect. Is there any documentation explaining how to make a package npm 7 compliant ? We can't find much informations about packages broken by npm 7.

**Reply by MylesBorins** [reply]
Does `--force` work?

**Reply by ljharb** [reply]
I wonder if it’s related to it using an “install” lifecycle script instead of “postinstall”. Did the install script behavior change in npm 7?

**Reply by lemoun** [reply]
Thanks for your help. `--force` doesn't work neither. Here is the log:
```
npm WARN using --force Recommended protections disabled.
npm WARN cleanup Failed to remove some directories [
npm WARN cleanup   [
npm WARN cleanup     'C:\\Users\\m\\project\\test\\node_modules',
npm WARN cleanup     [Error: EBUSY: resource busy or locked, rmdir 'C:\Users\m\project\test\node_modules\opencv-build'] {
npm WARN cleanup       errno: -4082,
npm WARN cleanup       code: 'EBUSY',
npm WARN cleanup       syscall: 'rmdir',
npm WARN cleanup       path: 'C:\\Users\\m\\project\\test\\node_modules\\opencv-build'
npm WARN cleanup     }
npm WARN cleanup   ]
npm WARN cleanup ]
npm ERR! code 1
npm ERR! path C:\Users\m\project\test\node_modules\opencv4nodejs
npm ERR! command failed
npm ERR! command C:\WINDOWS\system32\cmd.exe /d /s /c node ./install/install.js
npm ERR! info install using lib dir: C:/Users/m/project/test/node_modules/opencv-build/opencv/build/lib/Release
npm ERR! C:\Users\m\project\test\node_modules\opencv4nodejs\install\install.js:37
npm ERR!   throw new Error('library dir does not exist: ' + libDir)
npm ERR!   ^
npm ERR!
npm ERR! Error: library dir does not exist: C:/Users/m/project/test/node_modules/opencv-build/opencv/build/lib/Release
npm ERR!     at Object.<anonymous> (C:\Users\m\project\test\node_modules\opencv4nodejs\install\install.js:37:9)
npm ERR!     at Module._compile (node:internal/modules/cjs/loader:1108:14)
npm ERR!     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
npm ERR!     at Module.load (node:internal/modules/cjs/loader:973:32)
npm ERR!     at Function.Module._load (node:internal/modules/cjs/loader:813:14)
npm ERR!     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:76:12)
npm ERR!     at node:internal/main/run_main_module:17:47
```

**Reply by lemoun** [reply]
What are the fixes needed to make this package npm 7 compliant ? Can you point us to any direction ?

**Reply by MylesBorins** [reply]
I'm taking a look at the project and will move convo to the issue tracker there

**Reply by TymurSvorovskyi** [reply]
@lemoun your error log tells you, that `npm install` requires to be able to delete stuff in node_modules, but node_modules folder is already busy with another process where you probably trying to delete folder manually.

so just finish manually deleting /node_modules folder and run `npm install --wHatEvEr-pArAmEtERS`

**Reply by lemoun** [reply]
thanks for your help @TymurSvorovskyi but there is no manual deletion going on. The problem seems that @npm7 runs install hook before opencv4nodejs dependencies are installed, [as explained here](https://github.com/justadudewhohacks/opencv4nodejs/issues/796#issuecomment-830707031)
